### ISSUE TITLE
Host a public test server

### ISSUE THREAD

**Issue by satoshinm** [satoshinm/NetCraft#15]
Host a server somewhere (not in my browser at least until https://github.com/satoshinm/NetCraft/issues/14) open to anyone, for helping test and providing a live working environment. 

fogleman has an official Craft server connectable with https://github.com/fogleman/Craft#client `/online craft.michaelfogleman.com` in native, but web will necessarily not be compatible without changes (plain TCP vs WebSockets). After https://github.com/satoshinm/NetCraft/issues/12 multiplayer support in web version.

**Comment by satoshinm** [comment]
Registered for netcraft.heroku.com, to do: do something useful with it. Python instance running server.py? But needs a native cdll for chunk generation, can heroku host it? And a webserver serving the emscripten compiled craft.html/craft.js?

**Comment by satoshinm** [comment]
https://aternos.org/en/ looks promising ("Minecraft servers. Free. Forever.", really?), consider hosting NetCraft through WebSandboxMC now that https://github.com/satoshinm/WebSandboxMC/issues/28 is resolved.

**Comment by satoshinm** [comment]
Requested Aternos to add WebSandboxMC: https://board.aternos.org/index.php?thread/10904-plugin-request-websandboxmc/ - but their concept seems to be running MC servers on-demand cheaply, different ports on same machine (they gave me chicken.aternos.org:47566, created a Glowstone 1.11.2 server, with a vhost pointing to said port), so they may not want to add plugins which open ports, but couldn't hurt to ask.

If rejected by Aternos, perhaps try investigating Heroku some more, but instead of Craft's Python server.py, with [Glowstone](https://github.com/GlowstoneMC/Glowstone/) + [WebSandboxMC](https://github.com/satoshinm/WebSandboxMC/) - can it run Java?

**Comment by satoshinm** [comment]
Heroku is looking more promising, they don't allow TCP routing, but do allow WebSockets. Found this someone made: https://github.com/jacobwgillespie/heroku-minecraft - it also can use AWS to store the world data. That repo was using a reverse WS proxy to allow playing through TCP, but with WebSandboxMC, not strictly necessary! It's already websocketed.

**Comment by satoshinm** [comment]
Starting with https://github.com/jacobwgillespie/heroku-minecraft, modified to use Glowstone: https://github.com/satoshinm/heroku-glowstone - it is able to start a server, hwmtc listens on a port, but heroku log --tail shows memory errors:

2017-04-20T05:18:42.955539+00:00 app[web.1]: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>05:18:42 [INFO] Successfully bound server to 0.0.0.0/0.0.0.0:25565.
2017-04-20T05:18:55.707513+00:00 heroku[router]: at=info method=GET path="/tunnel" host=netcraft.herokuapp.com request_id=7afb3d42-a834-4247-b01d-dec21ceeb184 fwd="161.202.72.139" dyno=web.1 connect=0ms service=121ms status=101 bytes=129 protocol=http
2017-04-20T05:18:56.591808+00:00 heroku[router]: at=info method=GET path="/tunnel" host=netcraft.herokuapp.com request_id=2ebedad9-a29e-40fd-99a3-63b7daea5cfd fwd="161.202.72.139" dyno=web.1 connect=2ms service=36ms status=101 bytes=129 protocol=http
2017-04-20T05:19:03.771809+00:00 heroku[web.1]: Process running mem=611M(119.4%)
2017-04-20T05:19:03.771809+00:00 heroku[web.1]: Error R14 (Memory quota exceeded)

and it keeps increasing:

2017-04-20T05:23:15.402663+00:00 heroku[web.1]: Process running mem=650M(127.0%)
2017-04-20T05:23:15.402775+00:00 heroku[web.1]: Error R14 (Memory quota exceeded)

Heroku says: https://devcenter.heroku.com/articles/ruby-memory-use

**Comment by satoshinm** [comment]
Actually, error R14 is only a warning about swapping. It is possible to connect to the Glowstone server over websockets via hwmtc and it runs fine. Next up, making NetCraft connectable to it using WebSandboxMC.

What to do about the TCP socket routing? It may be only possible (or practical, perhaps possible with some kind of websocket multiplexer service?) to expose only one service/port, at a time, if so I'd choose the WebSocketMC HTTP/WS port (4081), over the WS'ified MC protocol port (25565). The MC protocol may not even be needed here, except possibly for administration/debugging (or can I use the server console?)

**Comment by satoshinm** [comment]
Got plugin installation working, WebSandboxMC is detected and runs (note using -L to curl to follow redirects): https://github.com/satoshinm/heroku-glowstone/commit/a0b11962afba05c47f5bb8a3202503f23452930a

Now to redirect 4081 to 80... or listen on $PORT ($GS_HEROKU_SERVER_PORT) instead of 4081 so Heroku forwards it to port 80.

**Comment by satoshinm** [comment]
Added configuration to use Heroku's port for WebSandboxMC: https://github.com/satoshinm/heroku-glowstone/commit/e769df72028f4c0fc7ffe1d43bc2f50b7e22bcfd

Now it loads and connects: http://netcraft.herokuapp.com

but the terrain I see isn't very interesting, it happens to be all stationary water - will need to find a suitable seed with more interesting terrain:

![screen shot 2017-04-20 at 12 20 09 am](https://cloud.githubusercontent.com/assets/26856618/25218331/5d505d1e-255f-11e7-898b-bbdf8191be13.png)

**Comment by satoshinm** [comment]
Spawn seems to differ with Glowstone even with consistent seed (config/glowstone.yml, world, seed: '1'), maybe instead of letting it generate, ship a fixed world file. All it needs is a small 32x32x32 region. And once it is seeded, should be able to back it up somewhere and restore.

**Comment by satoshinm** [comment]
Went with a boring seed for now: https://github.com/satoshinm/heroku-glowstone/commit/542474d21d392f44742371277d856ff154984834 - plain grass, nothing too interesting. But it isn't too critical, as it can be built on. So I built... this is how far I got:

![screen shot 2017-04-20 at 12 47 33 am](https://cloud.githubusercontent.com/assets/26856618/25219385/30ea9aa6-2563-11e7-9984-6856bec58599.png)

before the server crashed, `heroku logs --tail`:

```
2017-04-20T07:47:24.089986+00:00 heroku[web.1]: Process running mem=1143M(223.2%)
2017-04-20T07:47:24.089986+00:00 heroku[web.1]: Error R15 (Memory quota vastly exceeded)
2017-04-20T07:47:24.089986+00:00 heroku[web.1]: Stopping process with SIGKILL
2017-04-20T07:47:24.299311+00:00 heroku[router]: at=info method=GET path="/craftws" host=netcraft.herokuapp.com request_id=23772f2a-89f8-4f1c-8bc7-b16438f5502d fwd="161.202.72.139" dyno=web.1 connect=0ms service=214603ms status=101 bytes=211 protocol=http
2017-04-20T07:47:24.308847+00:00 heroku[web.1]: Process exited with status 137
2017-04-20T07:47:24.320832+00:00 heroku[web.1]: State changed from up to crashed
2017-04-20T07:47:36.713848+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/craft.html.mem" host=netcraft.herokuapp.com request_id=f6c5e5d3-a6c5-41a5-
```

"Vastly exceeded" memory quota, well, it isn't wrong. Glowstone really requires 1.143 GB? -Xmx and -Xms are set to 1024 MB, but Heroku's free dyno limit seems to be 512 MB. Trying to lower it...

**Comment by satoshinm** [comment]
Lowered the JVM memory parameters, seems to last longer, built this:

![screen shot 2017-04-20 at 1 04 54 am](https://cloud.githubusercontent.com/assets/26856618/25220079/85e71d8e-2565-11e7-856e-ff0e1abc8945.png)

but it is not saved, so will be destroyed when the dyno exits. This is where the Amazon S3 sync task comes in: https://github.com/satoshinm/heroku-glowstone/blob/master/bin/sync.rb - but haven't set it up yet. Meanwhile, this site can be considered highly experimental for testing multiple user interactions in NetCraft, temporarily: http://netcraft.herokuapp.com

**Comment by satoshinm** [comment]
Updated to a better seed in https://github.com/satoshinm/heroku-glowstone/commit/9b5f2684bf89010bd2534e461a0806ac93388849 - spawn on top of some nice trees, and disable nether/end for memory:

![screen shot 2017-04-20 at 8 30 15 pm](https://cloud.githubusercontent.com/assets/26856618/25261769/a753f354-2608-11e7-898c-de0667d94218.png)

but a generated spawn isn't too interesting. If it is going to be reset on each relaunch, copy a prebuilt spawn? At least with some signs/lights.

**Comment by satoshinm** [comment]
Turns out it *is* possible: unbind port 25565, then set port: 25565 (etc.) in plugins/WebSandboxMC/config.yml. Here is a server hosted at Server,pro available at http://158.69.124.46:40993/ (temporarily, until it shuts down after inactivity):

![screen shot 2017-04-22 at 11 30 58 pm](https://cloud.githubusercontent.com/assets/26856618/25311368/eb4382ee-27b3-11e7-8c0c-8236671945a3.png)

**Comment by satoshinm** [comment]
It isn't the most reliable, and doesn't persist to disk, but there is now a test server online: http://netcraft.herokuapp.com/ - also helping test https://github.com/satoshinm/WebSandboxMC. Got it to run reasonably well on Heroku by slimming down the open source Glowstone server: https://github.com/satoshinm/Glowstone/pull/1 into "Glowstone Light", using this buildpack: https://github.com/satoshinm/heroku-glowstone. Public testing has already found some useful bugs, open to anyone to join:

http://netcraft.herokuapp.com/
